### PR TITLE
go_2-dev: init at dev.go2go branch

### PR DESCRIPTION
**Pull request by 3noch**
###### Motivation for this change

Allow people to use the go2 proposal easily from Nix.

I'm not familiar with the build for Go so I only did things that managed to get this branch to build. ccing Go team in case they want to help support this cause in a more official way: @c00w @cstrahan @Frostman @kalbasit @mdlayher @mic92 @orivej @rvolosatovs @zowoq 

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/compilers/go/2-dev.nix [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+
+  postBuild = ''
+    (cd src && ./make.bash)
+  '';
+
+  doCheck = stdenv.hostPlatform == stdenv.targetPlatform && !stdenv.isDarwin;
+
+  checkPhase = ''
+    runHook preCheck
+    (cd src && HOME=$TMPDIR GOCACHE=$TMPDIR/go-cache ./run.bash --no-rebuild)
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/go/2-dev.nix:222]:
```suggestion

   pushd src
   HOME=$TMPDIR GOCACHE=$TMPDIR/go-cache ./run.bash --no-rebuild
   popd

```

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -9525,6 +9525,13 @@ in
     buildPackages = buildPackages // { stdenv = gcc8Stdenv; };
   });
 
+  go_2-dev = callPackage ../development/compilers/go/2-dev.nix ({
+    inherit (darwin.apple_sdk.frameworks) Security Foundation;
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:9529]:
```suggestion
    inherit (darwin.apple_sdk.frameworks) Foundation Security;
```

**Comment by zowoq** [comment]
@SuperSandro2000 I'd rather that this remain a copy of the existing `go1.*.nix` files for easier diffing, formatting changes can/should be applied to all of them.

**Comment by 3noch** [comment]
@SuperSandro2000 Yes this is just a copy of other go.nix files. Happy to clean all of them up but that should be a separate PR I'd imagine. I can separate out the maintainer changes. I'm probably not a good *choice* of maintainer for this since I don't actively use Go any more.

File: pkgs/development/compilers/go/2-dev.nix [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+    ${optionalString (stdenv.buildPlatform != stdenv.targetPlatform) ''
+    # Independent from host/target, CC should produce code for the building system.
+    # We only set it when cross-compiling.
+    export CC=${buildPackages.stdenv.cc}/bin/cc
+    ''}
+    ulimit -a
+  '';
+
+  postBuild = ''
+    (cd src && ./make.bash)
```
[the remark this answers is not in the document]
Comment (3noch) [reply pkgs/development/compilers/go/2-dev.nix:215]: I actually prefer the original way here. But like I said this is a copy.

**Comment by c00w** [comment]
Looks fine to me - I vote to merge :) (Pedantically I might make a go 2 folder for this, but honestly the go folder despite being a bit messy has never gotten in my way, so I don't worry about it too much).

File: maintainers/maintainer-list.nix [hunk 4]
```diff
@@ -70,6 +70,12 @@
     githubId = 7414843;
     name = "Nicholas von Klitzing";
   };
+  _3noch = {
```
[the remark this answers is not in the document]
Comment (3noch) [reply maintainers/maintainer-list.nix:73]: Done.

**Comment by kalbasit** [comment]
@GrahamcOfBorg build go_2-dev

**Comment by kalbasit** [comment]
The commit 

> go: Add dev.go2go branch variant called 2-dev

Does not really follow CONTRIBUTING.md and should be something like this: `go_2-dev: init at ...`

Otherwise LGTM. I'll merge once the commit is fixed and the bot builds it.

**Comment by 3noch** [comment]
@kalbasit How's that?

**Comment by kalbasit** [comment]
nix-review result

```
1 package built:
go_2-dev

[nix-shell:~/.cache/nixpkgs-review/pr-107538-1]$ ./results/go_2-dev/bin/go version
go version 2 linux/amd64
```

**Comment by fishy** [comment]
With the merge of this (or at least I believe so, I'm new to nixpkgs), my `go-1.15.6` auto upgraded to `go-2` with `nix-channel --update && nix-env -u`.

Should we rename this one as something like `go2-2` to avoid that? Or what's the correct way to "pin" my go at `1.*`, but still keep updated?

**Comment by zowoq** [comment]
> Should we rename this one as something like `go2-2` to avoid that?

#108270

**Comment by 3noch** [comment]
Yikes! I had no idea the update functionality worked like that. So thankful you pointed that out.

**Comment by kalbasit** [comment]
> Yikes! I had no idea the update functionality worked like that. So thankful you pointed that out.

Me neither :scream: Nix has some really weird edge cases.

**Pull request merged**